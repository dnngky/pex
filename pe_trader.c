/*
 * comp2017 - assignment 3
 * dan nguyen
 * kngu7458
 */

#include "pe_trader.h"

volatile sig_atomic_t pending_msgs;

/*
Close RFD and WFD, and terminate with exit code EXITCD.
*/
int term(int rfd, int wfd, int exitcd)
{
    close(rfd);
    close(wfd);
    return exitcd;
}

/*
Read the earliest pending message, from the pipe specified by RFD,
into BUFFER of size SIZE. Return the number of characters read.
*/
ssize_t get_order(int rfd, char *buffer, size_t size)
{
    char c[1];
    ssize_t nread = 0;
    while (nread < size) {
        read(rfd, c, 1);
        buffer[nread] = c[0];
        nread++;
        if (c[0] == ';')
            break;
    }
    buffer[nread] = '\0';
    pending_msgs--;

    return nread;
}

/*
Update the number of pending messages. This is a signal handler,
to be used by sigaction.
*/
void default_handler(int signum)
{
    pending_msgs++;
}

int main(int argc, char **argv)
{
	if (argc < 1) {
        puts("ERROR: Not enough arguments");
        return EXIT_FAILURE;
    }
    int tid = atoi(argv[1]);
    
    // connect to fifos
    char rfifo[BUFFER_SIZE];
    char wfifo[BUFFER_SIZE];
    if (snprintf(rfifo, BUFFER_SIZE, FIFO_EXCHANGE, tid) < 0) {
        perror(SNPRINTF_ERR);
        return EXIT_FAILURE;
    }
    if (snprintf(wfifo, BUFFER_SIZE, FIFO_TRADER, tid) < 0) {
        perror(SNPRINTF_ERR);
        return EXIT_FAILURE;
    }
    // open fifos
    int rfd = open(rfifo, O_RDONLY);
    if (rfd < 0) {
        perror("(EXCHANGE) OPEN ERROR");
        return EXIT_FAILURE;
    }
    int wfd = open(wfifo, O_WRONLY);
    if (wfd < 0) {
        perror("(TRADER) OPEN ERROR");
        return EXIT_FAILURE;
    }
    // register signal handler
    struct sigaction sa = {
        .sa_flags = SA_RESTART,
        .sa_handler = default_handler
    };
    sigaction(SIGUSR1, &sa, NULL);
    pending_msgs = 0;

    ssize_t nread;
    char rbuffer[BUFFER_SIZE];
    char wbuffer[BUFFER_SIZE];
    char msgbuffer[BUFFER_SIZE];

    // retrieve market open message
    while (pending_msgs == 0) pause();
	if ((nread = get_order(rfd, rbuffer, BUFFER_SIZE)) < 0) {
        perror(READ_ERR);
		return term(rfd, wfd, 1);
    }
    if (strncmp(rbuffer, MARKET_OPEN, strlen(MARKET_OPEN)) != 0) {
        write(STDOUT_FILENO, MARKET_NOPEN_ERR, strlen(MARKET_NOPEN_ERR));
        return term(rfd, wfd, 1);
    }
    int order_id = 0;
    
    // event loop:
    while (order_id <= ORDER_ID_LIMIT)
    {
        // retrieve message from exchange
        while (pending_msgs == 0) pause();
		if ((nread = get_order(rfd, rbuffer, BUFFER_SIZE)) < 0) {
            perror(READ_ERR);
			return term(rfd, wfd, 1);
        }
        // if message is not a notification about a SELL order, continue
        if (strncmp(rbuffer, "MARKET SELL", strlen("MARKET SELL")) != 0)
            continue;
        
        // else, parse sell order
        char type[5];
        char product[PRODUCT_MAXLEN+1];
        int qty;
        int price;

        int nscanned = sscanf(rbuffer, MARKET_NOTIFY, type, product, &qty, &price);
        if (nscanned == EOF) {
            perror(SSCANF_ERR);
            return term(rfd, wfd, 1);
        }
        if (nscanned < 4) {
            if (snprintf(msgbuffer, BUFFER_SIZE, INVALID_MSG_ERR, "PE", "MARKET SELL [product] [qty] [price]", rbuffer) < 0)
                perror(SNPRINTF_ERR);
            else
                write(STDOUT_FILENO, msgbuffer, strlen(msgbuffer));
            return term(rfd, wfd, 1);
        }
        // if maximum quantity is exceeded, terminate
        if (qty > TRADER_QTY_LIMIT)
            return term(rfd, wfd, 1);
        
        // write buy order
        if (snprintf(wbuffer, BUFFER_SIZE, BUY, order_id, product, qty, price) < 0) {
            perror(SNPRINTF_ERR);
            return term(rfd, wfd, 1);
        }
        if (write(wfd, wbuffer, strlen(wbuffer)) < 0) {
            perror(WRITE_ERR);
            return term(rfd, wfd, 1);
        }
        while (1)
        {
            // signal to exchange until a signal is received
            do {
                if (kill(getppid(), SIGUSR1) < 0) {
                    perror(KILL_ERR);
                    return term(rfd, wfd, 1);
                }
                usleep(TIMEOUT);
            } while (pending_msgs == 0);

            // retrieve message from exchange
            if ((nread = get_order(rfd, rbuffer, BUFFER_SIZE)) < 0) {
                perror(READ_ERR);
                return term(rfd, wfd, 1);
            }
            if (snprintf(wbuffer, BUFFER_SIZE, ACCEPTED, order_id) < 0) {
                perror(SNPRINTF_ERR);
                return term(rfd, wfd, 1);
            }
            // if order has been accepted, continue to next order
            if (strncmp(rbuffer, wbuffer, strlen(wbuffer)) == 0)
                break;
        }
        order_id++;
    }
    return term(rfd, wfd, 0);
}
