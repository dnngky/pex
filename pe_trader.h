#ifndef PE_TRADER_H
#define PE_TRADER_H

#include "pe_common.h"

// Limits
#define TRADER_QTY_LIMIT 999
#define MAX_WAITTIME 2

// Error
#define MARKET_NOPEN_ERR "Error: Market has not been opened\n"
#define INVALID_MSG_ERR "[%s] Error: Expected '%s'; Received '%s'\n"
#define INVALID_CMD_ERR "ERROR: Invalid trader command\n"

// Commands
#define BUY "BUY %d %s %d %d;"
#define SELL "SELL %d %s %d %d;"
#define AMEND "AMEND %d %d %d;"
#define CANCEL "CANCEL %d;"

#endif
