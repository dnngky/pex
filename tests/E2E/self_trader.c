#include "../../pe_trader.h"

volatile sig_atomic_t pending_msgs;

/*
Close RFD and WFD, and terminate with exit code EXITCD.
*/
int term(int rfd, int wfd, int exitcd)
{
    close(rfd);
    close(wfd);
    return exitcd;
}

/*
Read the earliest pending message, from the pipe specified by RFD,
into BUFFER of size SIZE. Return the number of characters read.
*/
ssize_t get_order(int rfd, char *buffer, size_t size)
{
    char c[1];
    ssize_t nread = 0;
    while (nread < size) {
        read(rfd, c, 1);
        buffer[nread] = c[0];
        nread++;
        if (c[0] == ';')
            break;
    }
    buffer[nread] = '\0';
    pending_msgs--;

    return nread;
}

/*
Update the number of pending messages. This is a signal handler,
to be used by sigaction.
*/
void default_handler(int signum)
{
    pending_msgs++;
}

int main(int argc, char **argv)
{
	if (argc < 1) {
        puts("ERROR: Not enough arguments");
        return EXIT_FAILURE;
    }
    int tid = atoi(argv[1]);
    
    // connect to fifos
    char rfifo[BUFFER_SIZE];
    char wfifo[BUFFER_SIZE];
    if (snprintf(rfifo, BUFFER_SIZE, FIFO_EXCHANGE, tid) < 0) {
        perror(SNPRINTF_ERR);
        return EXIT_FAILURE;
    }
    if (snprintf(wfifo, BUFFER_SIZE, FIFO_TRADER, tid) < 0) {
        perror(SNPRINTF_ERR);
        return EXIT_FAILURE;
    }
    // open fifos
    int rfd = open(rfifo, O_RDONLY);
    if (rfd < 0) {
        perror("(EXCHANGE) OPEN ERROR");
        return EXIT_FAILURE;
    }
    int wfd = open(wfifo, O_WRONLY);
    if (wfd < 0) {
        perror("(TRADER) OPEN ERROR");
        return EXIT_FAILURE;
    }
    // register signal handler
    struct sigaction sa = {
        .sa_flags = SA_RESTART,
        .sa_handler = default_handler
    };
    sigaction(SIGUSR1, &sa, NULL);
    pending_msgs = 0;

    ssize_t nread;
    char rbuffer[BUFFER_SIZE];
    char wbuffer[BUFFER_SIZE];

    // wait for market to open
    while (pending_msgs == 0) pause();
	if ((nread = get_order(rfd, rbuffer, BUFFER_SIZE)) < 0) {
        perror(READ_ERR);
		return term(rfd, wfd, 1);
    }
    if (strncmp(rbuffer, MARKET_OPEN, strlen(MARKET_OPEN)) != 0) {
        write(STDOUT_FILENO, MARKET_NOPEN_ERR, strlen(MARKET_NOPEN_ERR));
        return term(rfd, wfd, 1);
    }
    FILE *fp;
    fp = fopen("./tests/E2E/self_trader.in", "r");

    // event loop:
    int order_id = 0;
    while (1)
    {
        if (fgets(wbuffer, BUFFER_SIZE, fp) == NULL) {
            perror(FGETS_ERR);
            return term(rfd, wfd, 1);
        }
        wbuffer[strlen(wbuffer)-1] = '\0';
        if (strncmp(wbuffer, "DISCONNECT;", strlen("DISCONNECT;")) == 0) {
            fclose(fp);
            return term(rfd, wfd, 0);
        }
        if (write(wfd, wbuffer, strlen(wbuffer)) < 0) {
            perror(WRITE_ERR);
            fclose(fp);
            return term(rfd, wfd, 1);
        }
        // check command type
        char *MARKET_MSG = NULL;
        if (strncmp(wbuffer, "BUY", strlen("BUY")) == 0)
            MARKET_MSG = "ACCEPTED";
        else if (strncmp(wbuffer, "SELL", strlen("SELL")) == 0)
            MARKET_MSG = "ACCEPTED";
        else if (strncmp(wbuffer, "AMEND", strlen("AMEND")) == 0)
            MARKET_MSG = "AMENDED";
        else if (strncmp(wbuffer, "CANCEL", strlen("CANCEL")) == 0)  
            MARKET_MSG = "CANCELLED";
        else {
            write(STDOUT_FILENO, INVALID_CMD_ERR, strlen(INVALID_CMD_ERR));
            return term(rfd, wfd, 1);
        }
        while (1)
        {
            // signal to exchange until a signal is received or max waittime is reached
            int waittime = 0;
            do {
                if (kill(getppid(), SIGUSR1) < 0) {
                    perror(KILL_ERR);
                    return term(rfd, wfd, 1);
                }
                waittime++;
                usleep(TIMEOUT);
            } while (pending_msgs == 0 && waittime < MAX_WAITTIME);

            // retrieve message from exchange
            if ((nread = get_order(rfd, rbuffer, BUFFER_SIZE)) < 0) {
                perror(READ_ERR);
                return term(rfd, wfd, 1);
            }
            // if order has been accepted, continue to next order
            if (strncmp(rbuffer, MARKET_MSG, strlen(MARKET_MSG)) == 0)
                break;
        }
        order_id++;
    }
}
