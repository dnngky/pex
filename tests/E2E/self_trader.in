SELL 0 GPU 2 300;
BUY 1 Router 50 500;
BUY 2 GPU 2 500;
SELL 3 Router 1000 499;
DISCONNECT;
