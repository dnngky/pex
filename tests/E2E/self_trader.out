[PEX] Starting
[PEX] Trading 2 products: GPU Router
[PEX] Created FIFO /tmp/pe_exchange_0
[PEX] Created FIFO /tmp/pe_trader_0
[PEX] Starting trader 0 (./tests/E2E/self_trader)
[PEX] Connected to /tmp/pe_exchange_0
[PEX] Connected to /tmp/pe_trader_0
[PEX] [T0] Parsing command: <SELL 0 GPU 2 300>
[PEX]	--ORDERBOOK--
[PEX]	Product: GPU; Buy levels: 0; Sell levels: 1
[PEX]		SELL 2 @ $300 (1 order)
[PEX]	Product: Router; Buy levels: 0; Sell levels: 0
[PEX]	--POSITIONS--
[PEX]	Trader 0: GPU 0 ($0), Router 0 ($0)
[PEX] [T0] Parsing command: <BUY 1 Router 50 500>
[PEX]	--ORDERBOOK--
[PEX]	Product: GPU; Buy levels: 0; Sell levels: 1
[PEX]		SELL 2 @ $300 (1 order)
[PEX]	Product: Router; Buy levels: 1; Sell levels: 0
[PEX]		BUY 50 @ $500 (1 order)
[PEX]	--POSITIONS--
[PEX]	Trader 0: GPU 0 ($0), Router 0 ($0)
[PEX] [T0] Parsing command: <BUY 2 GPU 2 500>
[PEX] Match: Order 0 [T0], New Order 2 [T0], value: $600, fee: $6.
[PEX]	--ORDERBOOK--
[PEX]	Product: GPU; Buy levels: 0; Sell levels: 0
[PEX]	Product: Router; Buy levels: 1; Sell levels: 0
[PEX]		BUY 50 @ $500 (1 order)
[PEX]	--POSITIONS--
[PEX]	Trader 0: GPU 0 ($-6), Router 0 ($0)
[PEX] [T0] Parsing command: <SELL 3 Router 1000 499>
[PEX] Match: Order 1 [T0], New Order 3 [T0], value: $25000, fee: $250.
[PEX]	--ORDERBOOK--
[PEX]	Product: GPU; Buy levels: 0; Sell levels: 0
[PEX]	Product: Router; Buy levels: 0; Sell levels: 1
[PEX]		SELL 950 @ $499 (1 order)
[PEX]	--POSITIONS--
[PEX]	Trader 0: GPU 0 ($-6), Router 0 ($-250)
[PEX] Trader 0 disconnected
[PEX] Trading completed
[PEX] Exchange fees collected: $256
